File: server/app/schemas/tasks.py
from __future__ import annotations

"""Pydantic schemas for Unified Task Management."""

from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator


# ─── Subtask ─────────────────────────────────────────────────────────────────

class SubtaskItem(BaseModel):
    """A single subtask/checklist item within a task."""
    id: str = Field(..., description="Unique identifier (UUID)")
    title: str = Field(..., min_length=1, max_length=256)
    description: str | None = None  # optional text field for subtask details
    assigned_user_id: int | None = None  # backward compatibility
    assigned_user_ids: list[int] = Field(default_factory=list)  # array of user IDs
    report_spec: dict[str, Any] = Field(default_factory=dict)
    status: str = Field("pending", pattern="^(pending|in_progress|done|skipped)$")
    completed_at: datetime | None = None


class SubtaskItemCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=256)
    description: str | None = None  # optional text field for subtask details
    assigned_user_id: int | None = None  # backward compatibility
    assigned_user_ids: list[int] = Field(default_factory=list)  # array of user IDs
    report_spec: dict[str, Any] = Field(default_factory=dict)


# ─── Report Template ─────────────────────────────────────────────────────────

class ReportTemplateField(BaseModel):
    """A single field in a structured report template."""
    key: str = Field(..., min_length=1, max_length=64)
    label: str = Field(..., min_length=1, max_length=128)
    type: str = Field(..., pattern="^(text|number|select|boolean|datetime|textarea)$")
    required: bool = False
    options: list[str] = Field(default_factory=list)  # for select type


class ReportTemplate(BaseModel):
    """Task-level report template: structured fields and/or rich HTML body."""

    mode: str = Field(default="structured", pattern="^(structured|rich)$")
    fields: list[ReportTemplateField] = Field(default_factory=list)
    body_html: str = Field(default="", max_length=65536)
    attachments: list[dict[str, Any]] = Field(default_factory=list)

    @field_validator("attachments", mode="before")
    @classmethod
    def _coerce_attachments(cls, v: Any) -> Any:
        if v is None:
            return []
        return v


# ─── Task ────────────────────────────────────────────────────────────────────

class TaskCreate(BaseModel):
    task_type: str = Field(..., pattern="^(specific|routine)$")
    title: str = Field(..., min_length=1, max_length=256)
    description: str = Field("", max_length=4096)
    priority: str = Field("normal", pattern="^(low|normal|high|critical)$")
    patient_id: int | None = None
    assigned_user_id: int | None = None
    assigned_user_ids: list[int] = Field(default_factory=list)
    assigned_role: str | None = None
    start_at: datetime | None = None  # optional datetime for task start time
    ends_at: datetime | None = None
    due_at: datetime | None = None
    subtasks: list[SubtaskItemCreate] = Field(default_factory=list)
    report_template: ReportTemplate = Field(default_factory=ReportTemplate)
    report_template_pending_attachment_ids: list[str] = Field(default_factory=list)
    shift_date: date | None = None  # for routine tasks
    is_active: bool = True


class TaskUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=256)
    description: str | None = None
    priority: str | None = Field(None, pattern="^(low|normal|high|critical)$")
    start_at: datetime | None = None  # optional datetime for task start time
    ends_at: datetime | None = None
    due_at: datetime | None = None
    status: str | None = Field(None, pattern="^(pending|in_progress|completed|cancelled|skipped)$")
    assigned_user_id: int | None = None
    assigned_user_ids: list[int] = Field(default_factory=list)
    assigned_role: str | None = None
    subtasks: list[SubtaskItem] | None = None
    is_active: bool | None = None


class TaskOut(BaseModel):
    id: int
    workspace_id: int
    task_type: str
    patient_id: int | None
    title: str
    description: str
    priority: str
    start_at: datetime | None  # optional datetime for task start time
    ends_at: datetime | None
    due_at: datetime | None
    status: str
    assigned_user_id: int | None
    assigned_user_ids: list[int] = Field(default_factory=list)
    assigned_role: str | None
    created_by_user_id: int | None
    completed_at: datetime | None
    subtasks: list[dict[str, Any]] = Field(default_factory=list)
    report_template: dict[str, Any] = Field(default_factory=dict)
    workflow_job_id: int | None
    shift_date: date | None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    # Enriched fields (added at query time)
    patient_name: str | None = None
    assigned_user_name: str | None = None
    created_by_user_name: str | None = None
    report_count: int = 0

    model_config = {"from_attributes": True}


# ─── Task Report ─────────────────────────────────────────────────────────────

class TaskReportCreate(BaseModel):
    report_data: dict[str, Any] = Field(default_factory=dict)
    notes: str = Field("", max_length=512)
    attachments: list[str] = Field(default_factory=list)


class TaskReportOut(BaseModel):
    id: int
    workspace_id: int
    task_id: int
    patient_id: int | None
    submitted_by_user_id: int
    report_data: dict[str, Any]
    notes: str
    attachments: list[str]
    submitted_at: datetime

    submitted_by_user_name: str | None = None

    model_config = {"from_attributes": True}


# ─── Bulk Operations ─────────────────────────────────────────────────────────

class TaskBulkResetRequest(BaseModel):
    shift_date: date | None = None  # defaults to today if None


class TaskBoardUserRow(BaseModel):
    """Per-user aggregate row for the task board."""
    user_id: int
    username: str
    display_name: str
    role: str
    total: int
    in_progress: int
    completed: int
    skipped: int
    pending: int
    percent_complete: float
    tasks: list[TaskOut]


class TaskBoardResponse(BaseModel):
    shift_date: date | None = None
    rows: list[TaskBoardUserRow]

File: frontend/types/tasks.ts
/* ─────────────────────────────────────────────────────────────────────────────
   Unified Task Management Types
   ───────────────────────────────────────────────────────────────────────────── */

export interface SubtaskReportSpec {
  body_html?: string;
  attachment_hints?: string[];
  /** Client-only pending upload IDs; stripped server-side and merged into attachments. */
  attachment_pending_ids?: string[];
  attachments?: Array<{
    id: string;
    filename: string;
    content_type?: string;
    byte_size?: number;
    storage_relpath?: string;
  }>;
}

export interface SubtaskItem {
  id: string;
  title: string;
  description?: string;
  assigned_user_id?: number;
  assigned_user_ids?: number[];
  report_spec?: SubtaskReportSpec;
  status: string;
  completed_at?: string;
}

export interface ReportTemplateField {
  key: string;
  label: string;
  type: string;
  required?: boolean;
  options?: string[];
}

export interface ReportTemplate {
  mode?: "structured" | "rich" | string | null;
  fields: ReportTemplateField[];
  body_html?: string | null;
  attachment_hints?: string[];
  attachments?: Array<{
    id: string;
    filename: string;
    content_type?: string;
    byte_size?: number;
    storage_relpath?: string;
  }>;
}

export interface TaskCreate {
  task_type: "specific" | "routine";
  title: string;
  description?: string;
  priority: "low" | "normal" | "high" | "critical";
  patient_id?: number;
  assigned_user_id?: number;
  assigned_user_ids?: number[];
  assigned_role?: string;
  /** ISO datetime — shown on workspace calendar */
  start_at?: string;
  /** ISO datetime — optional window end */
  ends_at?: string;
  due_at?: string;
  subtasks?: Array<{
    title: string;
    description?: string;
    assigned_user_id?: number;
    assigned_user_ids?: number[];
    report_spec?: SubtaskReportSpec;
  }>;
  report_template?: ReportTemplate;
  report_template_pending_attachment_ids?: string[];
  shift_date?: string;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  priority?: string;
  status?: string;
  assigned_user_id?: number;
  assigned_user_ids?: number[];
  assigned_role?: string;
  start_at?: string;
  ends_at?: string;
  due_at?: string;
  subtasks?: SubtaskItem[];
  report_template?: ReportTemplate;
  is_active?: boolean;
}

export interface TaskOut {
  id: number;
  workspace_id: number;
  task_type: "specific" | "routine";
  patient_id?: number;
  title: string;
  description?: string;
  priority: string;
  start_at?: string;
  ends_at?: string;
  due_at?: string;
  status: string;
  assigned_user_id?: number;
  assigned_user_ids?: number[];
  assigned_role?: string;
  created_by_user_id?: number;
  completed_at?: string;
  subtasks: SubtaskItem[];
  report_template: ReportTemplate;
  workflow_job_id?: string;
  shift_date?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  patient_name?: string;
  assigned_user_name?: string;
  created_by_user_name?: string;
  report_count: number;
}

export interface TaskReportCreate {
  report_data: Record<string, unknown>;
  notes?: string;
  attachments?: string[];
}

export interface TaskReportOut {
  id: number;
  workspace_id: number;
  task_id: number;
  patient_id?: number;
  submitted_by_user_id: number;
  report_data: Record<string, unknown>;
  notes?: string;
  attachments: string[];
  submitted_at: string;
  submitted_by_user_name?: string;
}

export interface TaskBoardUserRow {
  user_id: number;
  username: string;
  display_name: string;
  role: string;
  total: number;
  in_progress: number;
  completed: number;
  skipped: number;
  pending: number;
  percent_complete: number;
  tasks: TaskOut[];
}

export interface TaskBoardResponse {
  shift_date?: string;
  rows: TaskBoardUserRow[];
}
